pkg/imagehandler/imagehandler.go

```diff
@@ -38,45 +85,84 @@ func (ie InvalidBaseImageError) Unwrap() error {
-	isoFile       *baseIso
-	initramfsFile *baseInitramfs
-	baseURL       *url.URL
-	keys          map[string]string
-	images        map[string]*imageFile
-	mu            *sync.Mutex
-	log           logr.Logger
+	isoFiles       map[string]*baseIso
+	initramfsFiles map[string]*baseInitramfs
+	baseURL        *url.URL
+	keys           map[string]string
+	images         map[string]*imageFile
+	mu             *sync.Mutex
+	log            logr.Logger
 }
 
 var _ ImageHandler = &imageFileSystem{}
 var _ http.FileSystem = &imageFileSystem{}
 
 type ImageHandler interface {
 	FileSystem() http.FileSystem
-	ServeImage(key string, ignitionContent []byte, initramfs, static bool) (string, error)
+	ServeImage(key string, arch string, ignitionContent []byte, initramfs, static bool) (string, error)
 	RemoveImage(key string)
 }
 
-func NewImageHandler(logger logr.Logger, isoFile, initramfsFile string, baseURL *url.URL) ImageHandler {
-	return &imageFileSystem{
-		log:           logger,
-		isoFile:       newBaseIso(isoFile),
-		initramfsFile: newBaseInitramfs(initramfsFile),
-		baseURL:       baseURL,
-		keys:          map[string]string{},
-		images:        map[string]*imageFile{},
-		mu:            &sync.Mutex{},
+func NewImageHandler(logger logr.Logger, baseURL *url.URL) (ImageHandler, error) {
+	imageFiles, err := os.ReadDir(imageSharedDir)
+
+	if err != nil {
+		return &imageFileSystem{}, err
+	}
+
+	isoFiles := map[string]*baseIso{}
+	initramfsFiles := map[string]*baseInitramfs{}
+
+	logger.Info("reading image files", "dir", imageSharedDir, "len", len(imageFiles))
+	for _, imageFile := range imageFiles {
+		filename := imageFile.Name()
+
+		logger.Info("load image", "imageFile", imageFile.Name())
+
+		ironicImage, err := parseIronicImage(filename)
+		if err != nil {
+			logger.Info("failed to parse ironic image, continuing")
+			continue
+		}
+
+		logger.Info("image loaded", "filename", ironicImage.filename, "arch", ironicImage.arch, "iso", ironicImage.iso, "initramfs", ironicImage.initramfs)
+
+		if ironicImage.iso {
+			isoFiles[ironicImage.arch] = newBaseIso(path.Join(imageSharedDir, filename))
+		}
+
+		if ironicImage.initramfs {
+			initramfsFiles[ironicImage.arch] = newBaseInitramfs(path.Join(imageSharedDir, filename))
+		}
 	}
+
+	return &imageFileSystem{
+		log:            logger,
+		isoFiles:       isoFiles,
+		initramfsFiles: initramfsFiles,
+		baseURL:        baseURL,
+		keys:           map[string]string{},
+		images:         map[string]*imageFile{},
+		mu:             &sync.Mutex{},
+	}, nil
 }
 
 func (f *imageFileSystem) FileSystem() http.FileSystem {
 	return f
 }
 
-func (f *imageFileSystem) getBaseImage(initramfs bool) baseFile {
+func (f *imageFileSystem) getBaseImage(arch string, initramfs bool) (baseFile, bool) {
+	if arch == "" {
+		arch = "host"
+	}
+
+	f.log.Info("getBaseImage", "arch", arch, "initramfs", initramfs)
 	if initramfs {
-		return f.initramfsFile
+		file, found := f.initramfsFiles[arch]
+		return file, found
```

Comment:
You need to implement the [`SupportsArchitecture()`](https://github.com/openshift/image-customization-controller/pull/144/files#diff-622ceeab03f53e5712a0dd562b04813ae0676fef344ef93982dc6b6cb04815b8R36-R38) function in `imageprovider/rhcos.go` in order to set the correct error when an architecture is not available.
If that is implemented correctly, there should be no case where the file is not found here.

Reply:
Done. I'll need to update the installer and cluster-baremetal-operator to add the new env var.

Reply:
I just realized we can get the architecture from the Golang runtime.